src/documentation/io/testing.md

```diff
@@ -108,11 +108,465 @@ This allows the I/O transform's unit tests to act as if they have a perfect netw
+1.  Tell Kubernetes to delete the resources specified in the Kubernetes scripts:
+    1.  JDBC: `kubectl delete -f ./test-infra/kubernetes/postgres/postgres.yml`
+    1.  Elasticsearch: `bash ./test-infra/kubernetes/elasticsearch/teardown.sh`
+
+
+### Implementing Integration Tests {#implementing-integration-tests}
+
+There are three components necessary to implement an integration test:
+*   **Test code**: the code that does the actual testing: interacting with the I/O transform, reading and writing data, and verifying the data.
+*   **Kubernetes scripts**: a Kubernetes script that sets up the data store that will be used by the test code.
+*   **Integrate with PerfKit Benchmarker using io-it-suite**: this allows users to easily invoke perfkit, creating the Kubernetes resources and running the test code.
+
+These three pieces are discussed in detail below.
+
+
+#### Test Code {#test-code}
+
+These are the conventions used by integration testing code:
+*   **Your test should use pipeline options to receive connection information.**
+    *   For Java, there is a shared pipeline options object in the io/common directory. This means that if there are two tests for the same data store (e.g. for `Elasticsearch` and the `HadoopInputFormatIO` tests), those tests share the same pipeline options.
+*   **Generate test data programmatically and parameterize the amount of data used for testing.**
+    *   For Java, `CountingInput` + `TestRow` can be combined to generate deterministic test data at any scale.
+*   **Use a write then read style for your tests.**
+    *   In a single `Test`, run a pipeline to do a write using your I/O transform, then run another pipeline to do a read using your I/O transform.
+    *   The only verification of the data should be the result from the read. Don't validate the data written to the database in any other way.
+    *   Validate the actual contents of all rows in an efficient manner. An easy way to do this is by taking a hash of the rows and combining them. `HashingFn` can help make this simple, and `TestRow` has pre-computed hashes.
+    *   For easy debugging, use `PAssert`'s `containsInAnyOrder` to validate the contents of a subset of all rows.
+*   **Tests should assume they may be run multiple times and/or simultaneously on the same database instance.**
+    *   Clean up test data: do this in an `@afterClass` to ensure it runs.
+    *   Use unique table names per run (timestamps are an easy way to do this) and per-method where appropriate.
+
+An end to end example of these principles can be found in [JdbcIOIT](https://github.com/ssisk/beam/blob/jdbc-it-perf/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java).
+
+If you'd like to implement Python I/O integration tests or integration tests for unbounded or eventually consistent data stores, please contact the Beam dev@ mailing list for more information.
```

Comment:
Recommend moving this to top of IT section as a note, perhaps with initial sentence around not supported/welcoming contributions in this area/whatever seems appropriate